### PR TITLE
Apim 4341 - List subscriptions

### PR DESCRIPTION
**Pull request by jansobczynski**
## Issue

https://gravitee.atlassian.net/browse/APIM-4341

## Description

Subscription list

## Additional context
Screenshot:
Screenshot 2024-05-21 at 2 50 41 PM

---

📚&nbsp;&nbsp;View the storybook of this branch [here](https://612657caa8e859003a8a6430-nxzauxqazc.chromatic.com)

### CODE REVIEW

File: gravitee-apim-portal-webui-next/src/app/api-details/api-tab-subscriptions/api-tab-subscriptions.component.html [hunk 1]
```diff
@@ -22,6 +22,29 @@
...
-    <p>Coming soon</p>
+    <table mat-table [dataSource]="subscriptions" class="api-tab-subscriptions__table">
+      <ng-container matColumnDef="application">
+        <th mat-header-cell *matHeaderCellDef i18n="@@subscriptionTableColumnApplication">Application</th>
+        <td mat-cell *matCellDef="let element">{{ element.application }}</td>
+      </ng-container>
+
+      <ng-container matColumnDef="plan">
+        <th mat-header-cell *matHeaderCellDef i18n="@@subscriptionTableColumnPlan">Plan</th>
+        <td mat-cell *matCellDef="let element">{{ subscriptionsPlanName(element.plan) }}</td>
```
Comment (jourdiw) [inline gravitee-apim-portal-webui-next/src/app/api-details/api-tab-subscriptions/api-tab-subscriptions.component.html]: Plan name should be titlecase I think. To be verified with Anthony + Audrey.

File: gravitee-apim-portal-webui-next/src/app/api-details/api-tab-subscriptions/api-tab-subscriptions.component.ts [hunk 2]
```diff
@@ -41,9 +45,18 @@ export class ApiTabSubscriptionsComponent implements OnInit {
     this.subscriptionsList$ = this.loadSubscriptions$();
   }
 
+  subscriptionsPlanName(plan: string) {
+    if (Object.hasOwn(this.subscriptionsMetadataList, plan)) {
```
Comment (jourdiw) [inline gravitee-apim-portal-webui-next/src/app/api-details/api-tab-subscriptions/api-tab-subscriptions.component.ts]: Every subscription will have a plan associated and every plan must have a name. So if the metadata is unreliable (aka you have to include a solution for if the plan name), you should get a list of plans and then match the plan id used in the subscription with the appropriate plan to get the plan name.

File: gravitee-apim-portal-webui-next/src/app/api-details/api-tab-subscriptions/api-tab-subscriptions.component.html [hunk 3]
```diff
@@ -22,6 +22,29 @@
...
+      </ng-container>
+
+      <ng-container matColumnDef="status">
+        <th mat-header-cell *matHeaderCellDef i18n="@@subscriptionTableColumnStatus">Status</th>
+        <td mat-cell *matCellDef="let element">{{ element.status }}</td>
```
Comment (jourdiw) [inline gravitee-apim-portal-webui-next/src/app/api-details/api-tab-subscriptions/api-tab-subscriptions.component.html]: This should be title case as well

File: gravitee-apim-portal-webui-next/src/app/api-details/api-tab-subscriptions/api-tab-subscriptions.component.scss [hunk 4]
```diff
@@ -20,3 +20,17 @@
...
+  overflow: hidden;
+  border: 1px solid #e6e0ec;
+  border-radius: 16px;
+
+  tr.mat-mdc-row.mdc-data-table__row {
+    background: #fff;
+  }
+
+  th.mat-mdc-header-cell {
+    font-weight: 600;
```
Comment (jourdiw) [inline gravitee-apim-portal-webui-next/src/app/api-details/api-tab-subscriptions/api-tab-subscriptions.component.scss]: Use one of the classes in `m3-adapter` to change the font style.

File: gravitee-apim-portal-webui-next/src/app/api-details/api-tab-subscriptions/api-tab-subscriptions.component.scss [hunk 5]
```diff
@@ -20,3 +20,17 @@
   justify-content: center;
   padding: 20px 0;
 }
+
+.api-tab-subscriptions__table {
+  overflow: hidden;
+  border: 1px solid #e6e0ec;
```
Comment (jourdiw) [inline gravitee-apim-portal-webui-next/src/app/api-details/api-tab-subscriptions/api-tab-subscriptions.component.scss]: Hex codes shouldn't be used directly... reference the palette with a `color-mix`.

File: gravitee-apim-portal-webui-next/src/app/api-details/api-tab-subscriptions/api-tab-subscriptions.component.scss [hunk 6]
```diff
@@ -20,3 +20,17 @@
...
+  border-radius: 16px;
+
+  tr.mat-mdc-row.mdc-data-table__row {
```
Comment (jourdiw) [inline gravitee-apim-portal-webui-next/src/app/api-details/api-tab-subscriptions/api-tab-subscriptions.component.scss]: Avoid using classes specific to material design. They can change names which will create style regressions in the future.

**Review by jourdiw — COMMENTED** [review]
Should add tests as well that the correct information is displayed.

File: gravitee-apim-portal-webui-next/src/app/api-details/api-tab-subscriptions/api-tab-subscriptions.component.html (hunk 3 above)
Comment (jansobczynski) [reply gravitee-apim-portal-webui-next/src/app/api-details/api-tab-subscriptions/api-tab-subscriptions.component.html]: Done

File: gravitee-apim-portal-webui-next/src/app/api-details/api-tab-subscriptions/api-tab-subscriptions.component.scss (hunk 4 above)
Comment (jansobczynski) [reply gravitee-apim-portal-webui-next/src/app/api-details/api-tab-subscriptions/api-tab-subscriptions.component.scss]: Done

File: gravitee-apim-portal-webui-next/src/app/api-details/api-tab-subscriptions/api-tab-subscriptions.component.ts (hunk 2 above)
Comment (jansobczynski) [reply gravitee-apim-portal-webui-next/src/app/api-details/api-tab-subscriptions/api-tab-subscriptions.component.ts]: I confirmed that all necessary data is in metadata object

File: gravitee-apim-portal-webui-next/src/app/api-details/api-tab-subscriptions/api-tab-subscriptions.component.scss (hunk 5 above)
Comment (jansobczynski) [reply gravitee-apim-portal-webui-next/src/app/api-details/api-tab-subscriptions/api-tab-subscriptions.component.scss]: Done

File: gravitee-apim-portal-webui-next/src/app/api-details/api-tab-subscriptions/api-tab-subscriptions.component.scss (hunk 6 above)
Comment (jansobczynski) [reply gravitee-apim-portal-webui-next/src/app/api-details/api-tab-subscriptions/api-tab-subscriptions.component.scss]: Done

**Comment by jansobczynski** [comment]
Tests added

File: gravitee-apim-portal-webui-next/src/app/api-details/api-tab-subscriptions/api-tab-subscriptions.component.html [hunk 7]
```diff
@@ -22,6 +22,29 @@
       <p i18n="@@subscribeToApi">Subscribe to our API and your subscription will show up here.</p>
     </div>
   } @else {
-    <p>Coming soon</p>
+    <table mat-table [dataSource]="subscriptions" class="api-tab-subscriptions__table">
+      <ng-container matColumnDef="application">
+        <th mat-header-cell *matHeaderCellDef i18n="@@subscriptionTableColumnApplication" class="m3-title-medium">Application</th>
+        <td mat-cell *matCellDef="let element">{{ retrieveMetadataName(element.application) }}</td>
```
Comment (jourdiw) [inline gravitee-apim-portal-webui-next/src/app/api-details/api-tab-subscriptions/api-tab-subscriptions.component.html]: Methods should be avoided in the html rendering because they are called incessantly, and in this case, the value will not be changing once rendered. Instead of doing the mapping here, move the mapping to the component after the service returns the Subscription response.

File: gravitee-apim-portal-webui-next/src/app/api-details/api-tab-subscriptions/api-tab-subscriptions.component.scss [hunk 8]
```diff
@@ -13,10 +13,23 @@
...
   padding: 20px 0;
 }
+
+.api-tab-subscriptions__table {
+  overflow: hidden;
+  border: 1px solid color-mix(in srgb, theme.$secondary-main-color 35%, transparent);
+  border-radius: 16px;
+
+  &-row {
+    background: #fff;
```
Comment (jourdiw) [inline gravitee-apim-portal-webui-next/src/app/api-details/api-tab-subscriptions/api-tab-subscriptions.component.scss]: Instead of using hex-code, use a variable. You can reference the mat card background color so that it matches. If we ever change to dark-mode, it will follow the same choice as the mat-cards.

File: gravitee-apim-portal-webui-next/src/app/api-details/api-tab-subscriptions/api-tab-subscriptions.component.spec.ts [hunk 9]
```diff
@@ -47,6 +52,24 @@ describe('ApiTabSubscriptionsComponent', () => {
...
 
+  describe('populated subscription list', () => {
+    beforeEach(() => {
+      expectSubscriptionList(fakeSubscriptionResponse(), 'testId');
+    });
+
+    it('should show subscription list', async () => {
+      const subscriptionTable = await harnessLoader.getHarness(MatTableHarness.with({ selector: '.api-tab-subscriptions__table' }));
+      expect(subscriptionTable).toBeTruthy();
+      console.log(3, subscriptionTable);
```
Comment (jourdiw) [inline gravitee-apim-portal-webui-next/src/app/api-details/api-tab-subscriptions/api-tab-subscriptions.component.spec.ts]: Remove console.log

File: gravitee-apim-portal-webui-next/src/entities/subscription/subscription.fixture.ts [hunk 10]
```diff
@@ -50,6 +50,11 @@ export function fakeSubscriptionResponse(
     links: {
       self: 'test',
     },
+    metadata: {
```
Comment (jourdiw) [inline gravitee-apim-portal-webui-next/src/entities/subscription/subscription.fixture.ts:53]: Since the application id should always exist with the name, it should be added here to reflect in the tests.

File: gravitee-apim-portal-webui-next/src/app/api-details/api-tab-subscriptions/api-tab-subscriptions.component.ts [hunk 11]
```diff
@@ -33,17 +37,28 @@ export class ApiTabSubscriptionsComponent implements OnInit {
...
   private subscriptionService = inject(SubscriptionService);
 
-  constructor() {}
+  constructor(private capitalizeFirstPipe: CapitalizeFirstPipe) {}
 
   ngOnInit(): void {
     this.subscriptionsList$ = this.loadSubscriptions$();
   }
 
+  retrieveMetadataName(plan: string) {
```
Comment (jourdiw) [inline gravitee-apim-portal-webui-next/src/app/api-details/api-tab-subscriptions/api-tab-subscriptions.component.ts]: It's a bit confusing that the parameter is named `plan` but you are using it for applications as well. Why not rename it to `id`?

File: gravitee-apim-portal-webui-next/src/pipe/capitalize-first.pipe.ts [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+ * limitations under the License.
+ */
+import { Pipe, PipeTransform } from '@angular/core';
+
+@Pipe({
+  name: 'capitalizeFirst',
+  standalone: true,
+})
+export class CapitalizeFirstPipe implements PipeTransform {
+  transform(value: string): string {
```
Comment (jourdiw) [inline gravitee-apim-portal-webui-next/src/pipe/capitalize-first.pipe.ts:23]: Maybe add a test that this works for empty strings `''` and one-letter strings, aka `'a'`?

**Review by jourdiw — CHANGES_REQUESTED** [review]
When all the changes are done, update the screenshot in the PR description to represent what you have developed. 👍

File: gravitee-apim-portal-webui-next/src/app/api-details/api-tab-subscriptions/api-tab-subscriptions.component.html (hunk 7 above)
Comment (jansobczynski) [reply gravitee-apim-portal-webui-next/src/app/api-details/api-tab-subscriptions/api-tab-subscriptions.component.html]: Done

File: gravitee-apim-portal-webui-next/src/app/api-details/api-tab-subscriptions/api-tab-subscriptions.component.scss (hunk 8 above)
Comment (jansobczynski) [reply gravitee-apim-portal-webui-next/src/app/api-details/api-tab-subscriptions/api-tab-subscriptions.component.scss]: Done

File: gravitee-apim-portal-webui-next/src/app/api-details/api-tab-subscriptions/api-tab-subscriptions.component.spec.ts (hunk 9 above)
Comment (jansobczynski) [reply gravitee-apim-portal-webui-next/src/app/api-details/api-tab-subscriptions/api-tab-subscriptions.component.spec.ts]: Done

File: gravitee-apim-portal-webui-next/src/pipe/capitalize-first.pipe.ts (hunk 12 above)
Comment (jansobczynski) [reply gravitee-apim-portal-webui-next/src/pipe/capitalize-first.pipe.ts:23]: Done

File: gravitee-apim-portal-webui-next/src/app/api-details/api-tab-subscriptions/api-tab-subscriptions.component.ts (hunk 11 above)
Comment (jansobczynski) [reply gravitee-apim-portal-webui-next/src/app/api-details/api-tab-subscriptions/api-tab-subscriptions.component.ts]: Done

File: gravitee-apim-portal-webui-next/src/entities/subscription/subscription.fixture.ts (hunk 10 above)
Comment (jansobczynski) [reply gravitee-apim-portal-webui-next/src/entities/subscription/subscription.fixture.ts:53]: Done

**Comment by jansobczynski** [comment]
Screenshot update:
Screenshot 2024-05-24 at 11 16 39 AM
I added one extra ui change, moved arrows to the right to match figma design

**Review by jourdiw — APPROVED** [review]
